src/Core/src/Hosting/ImageSources/ImageSourceToImageSourceServiceTypeMapping.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Maui.Hosting
+{
+	internal sealed class ImageSourceToImageSourceServiceTypeMapping
+	{
+		private static readonly ConcurrentDictionary<IImageSourceServiceCollection, ImageSourceToImageSourceServiceTypeMapping> s_instances = new();
+
+		internal static ImageSourceToImageSourceServiceTypeMapping GetInstance(IImageSourceServiceCollection collection) =>
+			s_instances.GetOrAdd(collection, static _ => new ImageSourceToImageSourceServiceTypeMapping());
+
+		private readonly Dictionary<Type, Type> _concreteTypeMapping = new();
+		private readonly Dictionary<Type, Type> _interfaceTypeMapping = new();
+
+		public void Add<TImageSource, TImageSourceService>()
+			where TImageSource : IImageSource
+			where TImageSourceService : class, IImageSourceService<TImageSource>
+		{
+			if (typeof(TImageSource).IsInterface)
+			{
+				_interfaceTypeMapping[typeof(TImageSource)] = typeof(TImageSourceService);
+			}
+			else
+			{
+				_concreteTypeMapping[typeof(TImageSource)] = typeof(TImageSourceService);
+			}
+		}
+
+		public Type FindImageSourceServiceType(Type type)
+		{
+			Debug.Assert(typeof(IImageSource).IsAssignableFrom(type));
+
+			if (_concreteTypeMapping.TryGetValue(type, out var exactImageSourceService)
+				|| _interfaceTypeMapping.TryGetValue(type, out exactImageSourceService))
+			{
+				return exactImageSourceService;
+			}
+
+			return FindImageSourceServiceMapping(type, _concreteTypeMapping)
+				?? FindImageSourceServiceMapping(type, _interfaceTypeMapping)
+				?? throw new InvalidOperationException($"Unable to find a {nameof(IImageSourceService)} corresponding to {type}. Please register a service for {type}.");
```

Comment:
If users decide to register their services using any of the `IServiceCollection` methods instead of `AddService` that may cause confusion
```suggestion
				?? throw new InvalidOperationException($"Unable to find a {nameof(IImageSourceService)} corresponding to {type}. Please register a service for {type} using `ImageSourceServiceCollectionExtensions.AddService`");
```

Reply:
Makes me wonder if we should have a test where we add services to the collection via `AddSingleton` instead of `AddService`.

Reply:
I added the test.